### PR TITLE
Support fp32 head for qwen and internlm models

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation

Support fp32 head for qwen and internlm models

## Modification

```python
from lmdeploy import pipeline, GenerationConfig, PytorchEngineConfig

if __name__ == '__main__':
    backend_config = PytorchEngineConfig(hf_overrides=dict(fp32_lm_head=True))
    model_path = 'Qwen/Qwen3-30B-A3B'
    pipe = pipeline(model_path, backend_config=backend_config)

    resps = pipe(['Hi.'])
    for res in resps:
        print(res)
```

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by grimoire** [comment]
Should we put `force_fp32_head` in BuildModelContext so we don't have to read and set args in every model?

File: lmdeploy/pytorch/models/utils/model.py [hunk 1]
```diff
@@ -106,13 +109,56 @@ def update_quant_config(cls, quant_config: QuantizationConfig):
...
+    def build_lm_head(self,
+                      hidden_size: int,
+                      vocab_size: int,
+                      bias: bool = False,
+                      dtype: Optional[torch.dtype] = None,
+                      device: Optional[torch.device] = None,
+                      **kwargs):
+        """Build LM Head."""
+        bm_ctx = get_build_model_context()
+        head_dtype = torch.float32 if bm_ctx.fp32_lm_head else dtype
```
Comment (lvhan028) [inline lmdeploy/pytorch/models/utils/model.py:140]: how about "self.head_dtype"?

File: lmdeploy/pytorch/models/utils/model.py [hunk 2]
```diff
@@ -106,13 +109,56 @@ def update_quant_config(cls, quant_config: QuantizationConfig):
...
+        """Compute logits of the model output."""
+        head_dtype = self.get_lm_head().weight.dtype
+        if hidden_states.dtype != head_dtype:
+            hidden_states = hidden_states.to(dtype=head_dtype)
+        hidden_states = self.get_lm_head()(hidden_states)
+        return hidden_states
+
+    def get_lm_head(self):
+        """Get lm_head."""
+        return self.lm_head
```
Comment (lvhan028) [inline lmdeploy/pytorch/models/utils/model.py]: What's the advantage of using ·get_lm_head` here? It seems we could just define `self.lm_head` directly in each model.

File: lmdeploy/pytorch/models/internlm2.py [hunk 3]
```diff
@@ -290,11 +291,7 @@ def __init__(self,
         # build Model
         self.model = InternLM2Model(config, dtype=dtype, device=device)
         # build lm_head
-        self.output = build_rowwise_linear(config.hidden_size,
-                                           config.vocab_size,
-                                           bias=False,
-                                           dtype=dtype,
-                                           device=device)
+        self.output = self.build_lm_head(config.hidden_size, config.vocab_size, bias=False, dtype=dtype, device=device)
```
Comment (lvhan028) [inline lmdeploy/pytorch/models/internlm2.py:294]: How about define `self.lm_head` like this `self.lm_head = self.output`?

File: lmdeploy/pytorch/models/internvl.py [hunk 4]
```diff
@@ -802,9 +802,9 @@ def forward(
                                                position_ids=position_ids,
                                                attn_metadata=attn_metadata)
 
-    def get_logits(self, hidden_states: torch.Tensor):
-        """Compute logits of the model output."""
-        return self.language_model.get_logits(hidden_states)
+    def get_lm_head(self):
+        """Get lm_head."""
+        return self.language_model.get_lm_head()
```
Comment (lvhan028) [inline lmdeploy/pytorch/models/internvl.py]: How about define `self.lm_head` in __init__?

File: lmdeploy/pytorch/models/utils/model.py (hunk 2 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/models/utils/model.py]: different model may have different names for this module.

Comment (lvhan028) [reply lmdeploy/pytorch/models/utils/model.py]: We can define `self_lm_head = the-different-name`, cann't we? For instance, `self.lm_head = self.output`

File: lmdeploy/pytorch/models/internvl.py (hunk 4 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/models/internvl.py]: It is a bit strange to have  `lm_head` from language_model to InternVLChatModel and it may be misleading when loading weights

File: lmdeploy/pytorch/models/utils/model.py (hunk 1 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/models/utils/model.py:140]: To reuse  line 116 `        head_dtype = self.get_lm_head().weight.dtype` ? I tried, but it may fail due to lm_head is on languange model class.

File: lmdeploy/pytorch/models/utils/model.py (hunk 2 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/models/utils/model.py]: Yes. We can. In that case, we need to add `self.lm_head = self.languange_model.lm_head` for vl models as well.  I'll change accordingly.
